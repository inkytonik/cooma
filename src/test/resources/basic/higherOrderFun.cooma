{
    val x : Vector(Int) = [1,2,3]
    {
        x = predef.map(Int, x , Int, fun (x : Int) predef.Ints.add(x, 1)),
        y = predef.fold(Int, x, fun (l : Int, r : Int) predef.Ints.add(l, r), 0)
    }
}
