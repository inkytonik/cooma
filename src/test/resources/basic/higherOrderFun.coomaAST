Program(
    Return(
        Blk(
            BlkLet(
                Let(
                    Val(),
                    IdnDef(
                        "x"),
                    Some(
                        VecT(
                            IntT())),
                    VecLit(
                        VecElems(
                            Vector(
                                Num(
                                    1),
                                Num(
                                    2),
                                Num(
                                    3))))),
                Return(
                    Rec(
                        Vector(
                            Field(
                                "x",
                                App(
                                    Sel(
                                        Idn(
                                            IdnUse(
                                                "predef")),
                                        FieldUse(
                                            "map")),
                                    Vector(
                                        IntT(),
                                        Idn(
                                            IdnUse(
                                                "x")),
                                        IntT(),
                                        Fun(
                                            Arguments(
                                                Vector(
                                                    Argument(
                                                        IdnDef(
                                                            "x"),
                                                        IntT(),
                                                        None))),
                                            App(
                                                Sel(
                                                    Ints(),
                                                    FieldUse(
                                                        "add")),
                                                Vector(
                                                    Idn(
                                                        IdnUse(
                                                            "x")),
                                                    Num(
                                                        1))))))),
                            Field(
                                "y",
                                App(
                                    Sel(
                                        Idn(
                                            IdnUse(
                                                "predef")),
                                        FieldUse(
                                            "fold")),
                                    Vector(
                                        IntT(),
                                        Idn(
                                            IdnUse(
                                                "x")),
                                        Fun(
                                            Arguments(
                                                Vector(
                                                    Argument(
                                                        IdnDef(
                                                            "l"),
                                                        IntT(),
                                                        None),
                                                    Argument(
                                                        IdnDef(
                                                            "r"),
                                                        IntT(),
                                                        None))),
                                            App(
                                                Sel(
                                                    Ints(),
                                                    FieldUse(
                                                        "add")),
                                                Vector(
                                                    Idn(
                                                        IdnUse(
                                                            "l")),
                                                    Idn(
                                                        IdnUse(
                                                            "r"))))),
                                        Num(
                                            0)))))))))))
