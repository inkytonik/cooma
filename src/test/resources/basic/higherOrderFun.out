{ x = [2,3,4], y = 6 }
